Check files affected by task contribution Pull Requests

RafDevX opened the pull request:
Fixes #2583 

This pull request checks whether any contributions match the norm of what is expected of a task proposal/submission:

- exactly one file changed, named `README.md` and in a reasonable location; and
- PR description matches the README file (so that reviewers don't need to check both places)

ludvigch committed:
check files affected by task contribution

Co-authored-by: Rafael Oliveira <rmfseo@kth.se>

RafDevX committed:
enforce PR description to match task README

This ensures that task contribution proposal/submission content is
identical in the pull request description and the affected README.md
file, so that reviewers can have confidence that they need only check
one of them without comparing against the other.

ludvigch committed:
make check-pr-description fail if diff

RafDevX commented on .github/workflows/check_task.yml:
```diff
@@ -12,8 +10,82 @@ on:
+            exit 1
+          fi
+
+  check-canvas:
+    name: Check proposal is compatible with previous student task registrations
```
Unsure whether we should require our new check to pass before running the existing Python script.
```suggestion
  check-canvas:
    name: Check proposal is compatible with previous student task registrations
    needs: check-pr-description
```
For now, we have opted not to specify a dependency so that the jobs can run in parallel and be faster for "good PRs", which should be the majority in volume, even if "bad PRs" may suffer slightly since the last job is not aborted even if another one fails first. In any case, the scale is arguably negligible; all the checks are relatively quick.

algomaster99 replied:
I agree here. We can run them independently.

ludvigch committed:
update PR template with new constraints

algomaster99 commented on .github/workflows/check_task.yml:
```diff
@@ -12,8 +10,82 @@ on:
+        # we use wdiff (word-diff) instead of plain diff (line-based) to ensure
+        # that we only enforce *content* to be the same rather than specific
+        # file formatting that would otherwise be ignored when the markdown is
+        # rendered. e.g., wdiff allows the README to break lines at col 80 for
+        # readability while still having the PR description without mid-sentence
+        # breaks (since GitHub descriptions don't support them)
```
What do you mean here? It is unclear. The diff is based on sequence of words, right? 
```suggestion
        # rendered. the comparison takes place between *sequence of words*.
        # eg. "- task 3\n## Category" would be tokenized as ["-task 3", "###Category"]
```
Writing this suggestion, made me question if removes spaces between "-" and "task 3" and so on.

RafDevX replied:
Yes, wdiff checks sequence of words, but with a more liberal interpretation of words: anything separated by whitespace is a word, so in your example `- Task 3` is equivalent to `-    Task 3`, but different from `-Task 3` (since then `-Task` would be the word under comparison). This check is fine because `-Task` isn't valid markdown for a list.

ludvigch replied:
To be more clear, in the suggestion, `"- task 3\n## Category"` would be tokenized as `["-", "Task", "3", "##", "Category"]` as far as we understand. 

We chose to use wdiff to both workaround the issue of new lines being different in the README and PR description because of Github's automatic formatting of the width as well as to provide students with more  flexibility in formatting with whitespace. At the same time the content is confirmed to be the same and the rendering is probably identical in both cases.

algomaster99 replied:
That's clear thanks!

algomaster99 commented on .github/workflows/check_task.yml:
```diff
@@ -12,8 +10,82 @@ on:
+
+          echo "affected_file=$FILES" >> "$GITHUB_OUTPUT"
+
+  check-pr-description:
+    name: Check whether Pull Request description matches affected file
+    needs: check-affected-files
+    runs-on: ubuntu-latest
+    steps:
```
This should trigger on description edit as well.

RafDevX replied:
Fixed in bd6321d30fb564ffed6b567ef5e98e8ea01846c1

algomaster99 reviewed (commented):
Thank you so much! Sorry for the delay in review. Left some questions/comments.

algomaster99 commented:
@RafDevX @ludvigch could you please have a look at my comments?

RafDevX commented:
Sorry, it's exam season - I'll take a look ASAP today or tomorrow.

algomaster99 commented:
No problem! Thanks for letting me know!

RafDevX committed:
trigger check_task workflow on PR edit

Since we check that the PR description matches the affected README file,
we need to repeat this check if the description ever changes.

Note that the GitHub default for if no types are specified is "opened",
"reopened", and "synchronize" -- so this commit only adds "edited", in
practice.

algomaster99 commented on .github/workflows/check_task.yml:
```diff
@@ -1,19 +1,96 @@
   pull_request_target:
+    types:
+      - opened
+      - reopened # good to check canvas again to see if eligibility changed
```
That's a good idea. I think `check-canvas` will take care of it, right?

algomaster99 reviewed (approved):
Looks good to me! Haven't tested it, but should work :)
